tiledb/api/src/query/subarray.rs

```diff
@@ -33,6 +39,121 @@ pub struct Subarray<'ctx> {
+                    let mut end_size: u64 = 0;
+                    ctx.capi_return(unsafe {
+                        ffi::tiledb_subarray_get_range_var_size(
+                            c_ctx,
+                            c_subarray,
+                            dim_idx,
+                            rng_idx,
+                            &mut start_size,
+                            &mut end_size,
+                        )
+                    })?;
+
+                    let start =
+                        vec![0u8; start_size as usize].into_boxed_slice();
+                    let end = vec![0u8; end_size as usize].into_boxed_slice();
+
+                    ctx.capi_return(unsafe {
+                        ffi::tiledb_subarray_get_range_var(
+                            c_ctx,
+                            c_subarray,
+                            dim_idx,
+                            rng_idx,
+                            start.as_ptr() as *mut std::ffi::c_void,
+                            end.as_ptr() as *mut std::ffi::c_void,
+                        )
+                    })?;
+
+                    let dtype = dim.datatype()?;
+                    let cvn = dim.cell_val_num()?;
+                    let range =
+                        TypedRange::from_slices(dtype, cvn, &start, &end)?
+                            .range;
+                    dim_ranges.push(range);
+                } else {
+                    let dtype = dim.datatype()?;
+                    let cvn = dim.cell_val_num()?;
+                    let size = match cvn {
+                        CellValNum::Fixed(cvn) => {
+                            cvn.get() as u64 * dtype.size()
+                        }
+                        // Unreachable becuase we're in !var_sized_dim
+                        CellValNum::Var => unreachable!(),
+                    };
+
+                    let start = vec![0u8; size as usize].into_boxed_slice();
+                    let end = vec![0u8; size as usize].into_boxed_slice();
+
+                    // Apparently stride exists in the API but isn't used.
+                    let mut stride: *const std::ffi::c_void = out_ptr!();
+
+                    ctx.capi_return(unsafe {
+                        ffi::tiledb_subarray_get_range(
+                            c_ctx,
+                            c_subarray,
+                            dim_idx,
+                            rng_idx,
+                            start.as_ptr() as *mut *const std::ffi::c_void,
+                            end.as_ptr() as *mut *const std::ffi::c_void,
+                            &mut stride,
+                        )
+                    })?;
+
+                    let range =
+                        TypedRange::from_slices(dtype, cvn, &start, &end)?
+                            .range;
+                    dim_ranges.push(range);
+                }
+            }
+
+            ranges.push(dim_ranges);
+        }
+
+        Ok(Vec::new())
```

Comment:
`Ok(ranges)`, no?
Did this not get caught in tests?

Reply:
Whoops. I hadn't added tests for this because I was waiting for the array strategy so that I'd just extend that. Then we shelved that for the time being. I'll throw in some local module tests to exercise this code so that I don't look so silly.

Reply:
[Add tests for Subarray::ranges()](https://github.com/TileDB-Inc/tiledb-rs/pull/72/commits/9a5186e97daee6201cb3a4b64b0bb66f4c115c35)